src/Compilers/Core/Portable/EncodedStringText.cs

```diff
@@ -170,30 +170,32 @@ private static byte[] TryGetByteArrayFromStream(Stream data)
         }
 
         /// <summary>
-        /// If the MemoryStream was created with publiclyVisible=true, then we can access its buffer
+        /// If the MemoryStream was created with publiclyVisible=true, then we can try to access its buffer
         /// directly and save allocations in StreamReader. The input MemoryStream is not closed on exit.
         /// </summary>
         /// <returns>True if a byte array could be created.</returns>
         private static bool TryGetByteArrayFromMemoryStream(MemoryStream data, out byte[] buffer)
         {
             Debug.Assert(data.Position == 0);
 
-            try
+            if (PortableShim.MemoryStream.TryGetBuffer != null)
             {
-                if (PortableShim.MemoryStream.GetBuffer != null)
+                ArraySegment<byte> bufferArraySegment;
+
+                if (PortableShim.MemoryStream.TryGetBuffer(data, out bufferArraySegment))
                 {
-                    buffer = (byte[])PortableShim.MemoryStream.GetBuffer.Invoke(data, null);
-                    return true;
+                    // since we can only return the whole array, make sure it's all used by the MemoryStream
+                    if (bufferArraySegment.Offset == 0 &&
+                        bufferArraySegment.Count == bufferArraySegment.Array.Length)
+                    {
```

Comment:
Yeah, you're right.

Though `Decode()` contains this check based on the stream length:

```c#
encoding.GetMaxCharCount((int)data.Length) < LargeObjectHeapLimitInChars
```

If the array is large but the part of it used by the `MemoryStream` is small, the change you propose would keep the large array around. Does this means that the LOH check should be repeated here, but this time checking the array length?